scrapyd_k8s/log_handler.py

```diff
@@ -0,0 +1,161 @@
+                                             prefix=f"{job_name}_logs_tmp_", suffix=".txt") as temp_logs:
+                temp_file_path = temp_logs.name
+                for line in w.stream(
+                    v1.read_namespaced_pod_log,
+                    name=job_name,
+                    namespace=namespace,
+                    follow=True,
+                    _preload_content=False
+                ):
+                    temp_logs.write(line + "\n")
+                    temp_logs.flush()
+
+                    if log_lines_counter == len(last_n_lines):
+                        log_file.write(line + "\n")
+                        log_file.flush()
+                    elif line == last_n_lines[log_lines_counter]:
+                        log_lines_counter += 1
+                    else:
+                        log_lines_counter = 0
+
+            if len(last_n_lines) > log_lines_counter:
+                concatenate_and_delete_files(log_file_path, temp_file_path)
+            else:
+                os.remove(temp_file_path)
+                logger.info(f"Removed temporary file '{temp_file_path}' after streaming logs for job '{job_name}'.")
+    except Exception as e:
+        logger.exception(f"Error streaming logs for job '{job_name}': {e}")
+
+def watch_pods(config):
+    object_storage_provider = LibcloudObjectStorage(config)
+    namespace = config.get_namespace()
+
+    num_lines_to_check = int(config.joblogs().get('num_lines_to_check'))
+    w = watch.Watch()
+    v1 = client.CoreV1Api()
+    try:
+        for event in w.stream(v1.list_namespaced_pod, namespace=namespace):
+            pod = event['object']
+            # check the labels in the docs
+            if pod.status.phase == 'Running' and pod.metadata.labels.get("org.scrapy.job_id"):
+                thread_name = "%s_%s" % (namespace, pod.metadata.name)
+                if (thread_name in watcher_threads
+                        and watcher_threads[thread_name] is not None
+                        and watcher_threads[thread_name].is_alive()):
+                    pass
+                else:
+                    watcher_threads[thread_name] = threading.Thread(
+                        target=stream_logs,
+                        kwargs={
+                            'job_name': pod.metadata.name,
+                            'namespace': namespace,
+                            'num_lines_to_check': num_lines_to_check
+                        }
+                    )
+                    watcher_threads[thread_name].start()
+            elif (pod.status.phase == 'Succeeded' or pod.status.phase == 'Failed') and pod.metadata.labels.get("org.scrapy.job_id"):
+                log_filename = pod_tmp_mapping.get(pod.metadata.name)
+                if log_filename is not None and os.path.isfile(log_filename) and os.path.getsize(log_filename) > 0:
+                    if object_storage_provider.is_local_file_uploaded(log_filename):
```

Comment:
I find the method name a bit confusing here, `object_exists` / `file_exists` or so would be clearer to me. As it is just checking whether an object is already present in the object storage or not.